### PR TITLE
[SPARK-11404] [SQL] Support for groupBy using column expressions

### PR DESCRIPTION
**Pull request by marmbrus**
This PR adds a new method `groupBy(cols: Column*)` to `Dataset` that allows users to group using column expressions instead of a lambda function.  Since the return type of these expressions is not known at compile time, we just set the key type as a generic `Row`.  If the user would like to work the key in a type-safe way, they can call `grouped.asKey[Type]`, which is also added in this PR.

``` scala
val ds = Seq(("a", 10), ("a", 20), ("b", 1), ("b", 2), ("c", 1)).toDS()
val grouped = ds.groupBy($"_1").asKey[String]
val agged = grouped.mapGroups { case (g, iter) =>
  Iterator((g, iter.map(_._2).sum))
}

agged.collect()

res0: Array(("a", 30), ("b", 3), ("c", 1))
```

### CODE REVIEW

**Comment by marmbrus** [comment]
@rxin @mateiz

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 1]
```diff
@@ -35,7 +35,8 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
   }
 
   test("as tuple") {
-    val data = Seq(("a", 1), ("b", 2)).toDF("a", "b")
+    val data = Seq(("a", 1), ("b", 2)).toDF("a", "c").select($"a", $"c".as("b"))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: why this change?

Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: Oh, i was live coding a demo at Spark summit and accidentally committed the change.  I'll revert.
